sql/api/src/main/scala/org/apache/spark/sql/types/DataType.scala

```diff
@@ -458,7 +458,7 @@ object DataType {
   private[sql] def equalsIgnoreCompatibleCollation(from: DataType, to: DataType): Boolean = {
     (from, to) match {
       // String types with possibly different collations are compatible.
-      case (_: StringType, _: StringType) => true
+      case (a: StringType, b: StringType) => a.constraint == b.constraint
```

Comment:
Don't know if this made some tests fail, but could you add tests for alter column with char/varchar to check what was behaviour before and after this change, so we do not introduce behavioural change?